Sources/ValidatingProperty.swift

```diff
@@ -0,0 +1,294 @@
+///
+/// ```
+/// let root = MutableProperty("Valid")
+/// let outer = root
+///   .validate { $0 == "Valid" ? nil : .outerInvalid }
+///
+/// outer.result.value        // `.success("Valid")
+///
+/// root.value = "🎃"
+/// outer.result.value        // `.failure("🎃", .outerInvalid)`
+/// ```
+public final class MutableValidatingProperty<Value, ValidationError: Swift.Error>: MutablePropertyProtocol {
+	private let getter: () -> Value
+	private let setter: (Value) -> Void
+
+	/// The result of the last attempted edit of the root property.
+	public let result: Property<ValidationResult<Value, ValidationError>>
+
+	/// The current value of the property.
+	///
+	/// The value could have failed the validation. Refer to `result` for the
+	/// latest validation result.
+	public var value: Value {
+		get { return getter() }
+		set { setter(newValue) }
+	}
+
+	public let producer: SignalProducer<Value, NoError>
+
+	public let signal: Signal<Value, NoError>
+
+	/// The lifetime of the editor.
+	public let lifetime: Lifetime
+
+	/// Create an `MutableValidatingProperty` that presents a mutable validating
+	/// view for an inner mutable property.
+	///
+	/// If `success` is returned by `validator`, the associated value would be
+	/// committed to `inner`. Otherwise, the failure would be exposed by the
+	/// `result` property of `self`.
+	///
+	/// - parameters:
+	///   - inner: The inner property which validated values are committed to.
+	///   - validator: The closure to invoke for any proposed value to `self`.
+	public init<Inner: ComposableMutablePropertyProtocol>(
+		_ inner: Inner,
+		_ validator: @escaping (Value) -> ValidatorOutput<Value, ValidationError>
+	) where Inner.Value == Value {
+		var mutesValueBackpropagation = false
+
+		getter = { inner.value }
+		producer = inner.producer
+		signal = inner.signal
+		lifetime = inner.lifetime
+
+		(result, setter) = inner.withValue { initial in
+			let mutableResult = MutableProperty(ValidationResult(initial, validator(initial)))
+
+			mutableResult <~ inner.signal
+				.filter { _ in !mutesValueBackpropagation }
+				.map { ValidationResult($0, validator($0)) }
+
+			return (Property(capturing: mutableResult), { input in
+				let writebackValue: Value? = mutableResult.modify { result in
+					result = ValidationResult(input, validator(input))
+					return result.value
+				}
+
+				if let value = writebackValue {
+					mutesValueBackpropagation = true
+					inner.value = value
+					mutesValueBackpropagation = false
+				}
+			})
+		}
+	}
+
+	/// Create an `MutableValidatingProperty` that validates mutations before
+	/// committing them.
+	///
+	/// If `success` is returned by `validator`, the associated value would be
+	/// committed to `inner`. Otherwise, the failure would be exposed by the
+	/// `result` property of `self`.
+	///
+	/// - parameters:
+	///   - initial: The initial value of the property. It is not required to
+	///              pass the validation as specified by `validator`.
```

Comment:
Consider a form field that requires at least three characters. Initially it has to be invalid.